### PR TITLE
generate random keys for use in tracker announces

### PR DESCRIPTION
**Pull request by arvidn**
keys are unique for each torrent and the listen interface they announce via

### CODE REVIEW

**Comment by arvidn** [comment]
@ssiloti I would appreciate if you would take a look. this is related to https://github.com/arvidn/libtorrent/issues/2255. I generate stable keys for the listen_sockets (based on the local address) and XOR them with the per-torrent key. Perhaps there are better ways to combine them but I figure it's not that big of a deal.

File: include/libtorrent/aux_/session_impl.hpp [hunk 1]
```diff
@@ -906,11 +912,6 @@ namespace aux {
 			// the peer id that is generated at the start of the session
 			peer_id m_peer_id;
 
-			// the key is an id that is used to identify the
-			// client with the tracker only. It is randomized
-			// at startup
-			std::uint32_t m_key = 0;
```
Comment (arvidn) [inline include/libtorrent/aux_/session_impl.hpp]: no session wide key anymore

File: include/libtorrent/aux_/session_impl.hpp [hunk 2]
```diff
@@ -198,6 +198,10 @@ namespace aux {
 		// which alias the listen_socket_t shared_ptr
 		std::shared_ptr<tcp::acceptor> sock;
 		std::shared_ptr<aux::session_udp_socket> udp_sock;
+
+		// the key is an id that is used to identify the
+		// client with the tracker only.
+		std::uint32_t tracker_key = 0;
```
Comment (arvidn) [inline include/libtorrent/aux_/session_impl.hpp]: the session wide keys are generated per listen socket now

File: include/libtorrent/session_handle.hpp [hunk 3]
```diff
@@ -586,10 +586,13 @@ namespace libtorrent {
 		// the peer ID is randomized per peer.
 		peer_id id() const;
 
+#ifndef TORRENT_NO_DEPRECATE
+		// deprecated in 1.2
 		// sets the key sent to trackers. If it's not set, it is initialized
 		// by libtorrent. The key may be used by the tracker to identify the
 		// peer potentially across you changing your IP.
 		void set_key(std::uint32_t key);
```
Comment (arvidn) [inline include/libtorrent/session_handle.hpp]: I'm not sure what people used this for anyway

File: src/session_impl.cpp [hunk 4]
```diff
@@ -1384,6 +1394,25 @@ namespace {
 			reopen_outgoing_sockets();
 	}
 
+	std::uint32_t session_impl::get_tracker_key(address const& iface) const
```
Comment (arvidn) [inline src/session_impl.cpp]: generates a (stable) key for a given listen socket address

File: src/session_impl.cpp [hunk 5]
```diff
@@ -1384,6 +1394,25 @@ namespace {
...
+	{
+		uintptr_t const ses = reinterpret_cast<uintptr_t>(this);
+		hasher h(reinterpret_cast<char const*>(&ses), sizeof(ses));
+		if (iface.is_v4())
+		{
+			auto const b = iface.to_v4().to_bytes();
+			h.update({reinterpret_cast<char const*>(b.data()), b.size()});
```
Comment (ssiloti) [inline src/session_impl.cpp]: I'm not sure why anyone would do it, but in theory there could be multiple listen sockets with the same address but different ports. To get unique keys for every listen socket in this case the `original_port` would also need to be incorporated into the hash.

**Comment by ssiloti** [comment]
Just one comment, otherwise lgtm

**Comment by arvidn** [comment]
I’m thinking of the case where the client changes listen port, but only has one interface say. It would be strange for the key to change.

**Pull request merged**